tests/ui/associated-type-bounds/duplicate-bound.rs

```diff
@@ -0,0 +1,240 @@
+where
+    T: Iterator<Item: Copy, Item: Copy>,
+{
+    f: ManuallyDrop<T>,
+}
+union UW3<T>
+where
+    T: Iterator<Item: 'static, Item: 'static>,
+{
+    f: ManuallyDrop<T>,
+}
+
+fn fi1<T: Iterator<Item: Copy, Item: Send>>() {}
+fn fi2<T: Iterator<Item: Copy, Item: Copy>>() {}
+fn fi3<T: Iterator<Item: 'static, Item: 'static>>() {}
+fn fw1<T>()
+where
+    T: Iterator<Item: Copy, Item: Send>,
+{
+}
+fn fw2<T>()
+where
+    T: Iterator<Item: Copy, Item: Copy>,
+{
+}
+fn fw3<T>()
+where
+    T: Iterator<Item: 'static, Item: 'static>,
+{
+}
+
+fn frpit1() -> impl Iterator<Item: Copy, Item: Send> {
+    iter::empty::<u32>()
+}
+fn frpit2() -> impl Iterator<Item: Copy, Item: Copy> {
+    iter::empty::<u32>()
+}
+fn frpit3() -> impl Iterator<Item: 'static, Item: 'static> {
+    iter::empty::<u32>()
+}
+fn fapit1(_: impl Iterator<Item: Copy, Item: Send>) {}
+fn fapit2(_: impl Iterator<Item: Copy, Item: Copy>) {}
+fn fapit3(_: impl Iterator<Item: 'static, Item: 'static>) {}
+
+type TAI1<T: Iterator<Item: Copy, Item: Send>> = T;
+type TAI2<T: Iterator<Item: Copy, Item: Copy>> = T;
+type TAI3<T: Iterator<Item: 'static, Item: 'static>> = T;
+type TAW1<T>
+where
+    T: Iterator<Item: Copy, Item: Send>,
+= T;
+type TAW2<T>
+where
+    T: Iterator<Item: Copy, Item: Copy>,
+= T;
+type TAW3<T>
+where
+    T: Iterator<Item: 'static, Item: 'static>,
+= T;
+
+trait TRI1<T: Iterator<Item: Copy, Item: Send>> {}
+trait TRI2<T: Iterator<Item: Copy, Item: Copy>> {}
+trait TRI3<T: Iterator<Item: 'static, Item: 'static>> {}
+trait TRS1: Iterator<Item: Copy, Item: Send> {}
+trait TRS2: Iterator<Item: Copy, Item: Copy> {}
+trait TRS3: Iterator<Item: 'static, Item: 'static> {}
+trait TRW1<T>
+where
+    T: Iterator<Item: Copy, Item: Send>,
+{
+}
+trait TRW2<T>
+where
+    T: Iterator<Item: Copy, Item: Copy>,
+{
+}
+trait TRW3<T>
+where
+    T: Iterator<Item: 'static, Item: 'static>,
+{
+}
+trait TRSW1
+where
+    Self: Iterator<Item: Copy, Item: Send>,
+{
+}
+trait TRSW2
+where
+    Self: Iterator<Item: Copy, Item: Copy>,
+{
+}
+trait TRSW3
+where
+    Self: Iterator<Item: 'static, Item: 'static>,
+{
+}
+trait TRA1 {
+    type A: Iterator<Item: Copy, Item: Send>;
+}
+trait TRA2 {
+    type A: Iterator<Item: Copy, Item: Copy>;
+}
+trait TRA3 {
+    type A: Iterator<Item: 'static, Item: 'static>;
+}
+
+trait Trait {
+    type Gat<T>;
+
+    const ASSOC: i32;
+
+    fn foo() -> impl Sized;
+}
+
+impl Trait for () {
+    type Gat<T> = ();
+
+    const ASSOC: i32 = 3;
+
+    fn foo() {}
+}
+
+trait Subtrait: Trait<Gat<u32> = u32, Gat<u64> = u64> {}
+
+fn f<T: Trait<Gat<i32> = (), Gat<i64> = ()>>() {
+    let _: T::Gat<i32> = ();
+    let _: T::Gat<i64> = ();
+}
+
+fn g<T: Trait<Gat<i32> = (), Gat<i64> = &'static str>>() {
+    let _: T::Gat<i32> = ();
+    let _: T::Gat<i64> = "";
+}
+
+fn uncallable(_: impl Iterator<Item = i32, Item = u32>) {}
+
+fn callable(_: impl Iterator<Item = i32, Item = i32>) {}
+
+fn uncallable_const(_: impl Trait<ASSOC = 3, ASSOC = 4>) {}
```

Comment:
It feels weird to me that we would accept such code without linting on it, defining functions/ADTs/whatever that can never have their where clauses satisfied doesn't seem like the kind of thing people would want to intentionally do.

I would like us to add some `unsatisfiable_bounds` lint to detect this kind of case and warn people on it :thinking: